### ISSUE TITLE
Fast primality test for safe prime candidates

### ISSUE THREAD

**Issue by maroonbells** [openssl/openssl#17956]
Related to closed issue: Fixed range of random produced in BN_is_prime_fasttest_ex() #6547

Does v3.0 no longer have an external function which can do a fast primality test for safe primes?

Now that BN_is_prime_ex() and BN_is_prime_fasttest_ex() and BN_check_prime() have become a wrapper around ossl_bn_check_prime, they can't perform a fast primality test anymore if it involves safe primes.

The use case is where a range of safe prime candidates has been sieved to eliminate many more composites than are eliminated by the limited size of openssl's trial divisions table of small primes.

For the remaining sieved candidates where the vast majority of them or both not-prime and not-safe-prime, they don't need any trial divisions because the sieve has already done that against both p and q, and more. The vast majority of these remaining candidates can be rejected by calling libcrypto-1_1.dll's BN_is_prime_ex() to test 'p' with 1 round of Miller-Rabin before testing the 'q' candidate. By having v3.0 enforcing a minimum of 128 rounds of Miller-Rabin against bit lengths above 2048, this is wasting a lot of time doing 127 rounds of Miller-Rabin against 'p' safe prime candidates where the vast majority of them will be rejected due to 1 round of Miller-Rabin being enough to determining that 'q' is composite.

Or, is there an available external .dll function which can perform 1 round of Miller-Rabin but isn't labeled as 'is prime', and it can do the equivalent job?

**Comment by t8m** [comment]
There is no such public function.

**Comment by maroonbells** [comment]
Well perhaps you might consider restoring this functionality which was taken away. At the very least there could be a function where there's a parameter that lets you choose whether to test 1 round against "q = input >> 1" or "p = input << 1 +1" before wasting time doing 128 rounds of MR against many numbers who are not a safe prime and/or Sophie Germain prime

**Comment by kroeckx** [comment]
Is there a reason you're not using OpenSSL to generate the safe prime?
As far as I know, it should do right thing, only sieve once, switch
between p and q for the MR rounds.

The reason for only providing the current functionallity is that it
prevents people from misusing it.

**Comment by t8m** [comment]
We could add a `BN_check_safe_prime()` function. That should not be easily misused.

**Comment by maroonbells** [comment]
The external interface does not allow defining the starting position for the search for a safe prime, which among other things prevents finding safe primes designed like RFC 7919/3526 where the top/bottom 64 bits are each 1's. The OpenSSL search for safe primes benchmarks much slower than it needs to be, possibly because the functionality is piggybacked into a function designed primarily to find RSA primes.

I cannot identify where OpenSSL is doing any sieving, which is different than doing trial divisions. Sieving is identifying a candidate being a composite multiple of a small prime 'r', and then using that knowledge to immediately cross off every r'th candidate across a large range, without needing to do any more divisions.

A search would do either sieving or trial divisions, not both, and I see OpenSSL doing trial divisions. Once a range has been sieved, it just does M/R against 100% of the remaining candidates. But even when a sieve has been done against both p and q - for all candidates where 'p' looks like a probable prime due to passing 1 M/R round, there's no need to do more than 1 M/R round against either p or q until both p and q have passed at least 1 M/R round.

My tests have found that, compared to doing an actual sieve, the number of trial divisions determined by calc_trial_divisions is not efficient for safe prime searches, and probably isn't efficient for RSA primes either, though the FIPS design for linking p and q together seems not to worry about speed of finding RSA primes, and is more concerned about a deterministic relationship between the 2 RSA primes. The lower the quantity of trial divisions, the greater is the number of candidates which must face M/R tests, which can greatly slow down the search for safe primes.

Among large numbers, only numbers like 12k+11 can ever be safe primes, so trial division by 2 and 3 don't ever help unless safe prime candidates are being tested at intervals of 4 instead of at intervals of 12. For the 12k+11 numbers, the portion of candidates not eliminated by trial division is (r-2)/r for each small prime 'r' used in a trial division, so the percentage of 12k+11 candidates needing to face an M-R test in OpenSSL is:

4.0075% 2048-bits  384 trials
3.0728% 3072-bits 1024 trials
3.0728% 4096-bits 1024 trials
2.6016% 6144-bits 2048 trials
2.6016% 8192-bits 2048 trials

Doing a sieve against 2048 small primes would be much faster than doing trial divisions against 2048 small primes because there only needs to be a single BIGNUM/r division, because from this you can calculate which is the 1st position, if any, within the range where 'p' is a multiple of 'r', and same for finding the position where 'q' is a multiple of 'r'. Then once you've identified a candidate that fails the sieve, you get to cross off every r'th candidate within the range for virtually free, just the cost of a uint32 addition to reach the spot in the sieve index where you cross off the position 'r' distance later.

Based on the density of Sophie Germain q's and the Twin Prime constant, the mean number of p=12k+11 candidates encountered in a search before finding an 'n' bit safe prime is approx ((n-1)*log(2))^2/(2*TPC)/6), so the mean number of candidates encountered before reaching a safe prime, and the number of M/R tests would be:

bits:
2048 254129 positions * 4.0075% = 10184 tested by M/R
3072 571977 positions * 3.0728% = 17576 tested by M/R
4096 1017014 positions * 3.0728% = 31251 tested by M/R
6144 2288655 positions * 2.6016% = 59543 tested by M/R
8192 4069051 positions * 2.6016% = 105863 tested by M/R

So, if OpenSSL were really sieving, the time spent sieving 1024 small primes against 2048 bits, instead of sieving against the 384 smallest r's, would be a fraction of the time saved by doing M/R against (4.0075 - 3.0728)/4.0075) = 23.3% fewer candidates.

Sieving against the first 100k primes would lower the candidates needing M/R testing down to 1.30%, but sieving is not the same as trial divisions. Sieving is much faster than trial division because there are no candidates having multiple trial divisions prior to facing M/R.

As a test, I compared the speed of v3.01 doing "openssl dhparam -noout 2048"
against c# code that's using random 2048-bit numbers as the starting position for a safe prime search, then the c# sieves p and q candidates against the smallest 20 million primes. Once the sieving step is done, there's no need for any trials, the c# simply does M/R against the 0.64% of positions not flagged by the sieve as composite.

Because I can't figure out from the OpenSSL display of dots and plus'es how many candidates have been tested, I can't accurately identify the speed at which OpenSSL processes the range of candidates. So, to make the test more fair, I tried to do an average, which used 27 tests and discarded the longest and slowest time as outliers, leaving 25 tests from which to take an average.

For OpenSSL 3.01, the time for finding 27 safe primes was 01:19:15 less fastest 3 secs less slowest 09:46 = 01:09:26 (69.5 minutes).

For the c# program, the total time for 27 safe primes was 26:54, less 28 secs for the fastest time, less slowest time of 2 mins 54sec = 00:23:32 (23.5 minutes)

OpenSSL was more than twice as slow. OpenSSL had 9 of 27 cases which were slower than the sieved program's slowest time of 2:54, and OpenSSL's average time was still slower than the sieved program's average even if i tossed all 9 of OpenSSL's times slower than 2:54.

When testing at 3072 bits, I had to lower the test count because OpenSSL's results were so slow that it would have taken way too long, and I instead had OpenSSL 3.01 repeat the test 12 times instead of 27, but still discard the fastest/slowest as outliers.

For the c# program, the time for 27 safe primes at 3072 bits was 01:38:59 less 42 secs for the fastest time to find a safe prime, less slowest time 13:01 = 01:25:16 for 25 tests. (85 minutes)

For OpenSSL, the time for 12 3072-bit safe primes was 4:09:14 hrs less fastest time 05:31, less slowest time 44:47 = 03:18:56 for 10 tests. (198 minutes) This was more than twice as long to find 40% as many safe primes.

The c# program called the 1.1.1L .dll to perform the M/R tests, so the only difference between the c# program and OpenSSL should have been the choice of sieving and trial division parameters.

At 2048 bits it took 28 seconds to sieve both p and q against the first 20 million primes greater than 3, and then it simply did M/R tests against the remaining 0.64% of 12k+11 candidates that were not eliminated by the sieve. At 3072 bits, the only difference from 2048-bits is that the sieve took 38 seconds instead of 28. Assuming that OpenSSL was doing M/R tests against the 3.0728% of 12k+11 candidates that passed the trial division against the smallest 1024 primes, that means the c# program was able to trade an extra 38 seconds of sieving 3072-bit candidates in exchange for having (3.0728-0.64)/3.0728 = 79.2% fewer M/R tests, and the longer the bit length, the more attractive this trade-off becomes.

**Comment by t8m** [comment]
Another option would be to submit your safe prime generation algorithm as a PR for master branch. We would most probably accept it.
